### PR TITLE
[NuGet Symbol Server] Gallery upload for Symbols package

### PR DESCRIPTION
**Pull request by shishirx34**
This PR adds support for the uploading symbols package to the gallery. 

I refactored a bit more code within the `PackagesController`, for convenience. Also added tests.

I will verify this branch on `Dev` env before merging it in. Also, added test cases for the new things. I have not refactored too much with regards to the `package` upload workflow itself, so as to not break the primary flow upload flow. @diverdan92 - was cool with the UI changes. If you have any UI related feedback, please feel free to tag @diverdan92 as well into the conversation.

I will add screenshots subsequently, when I validate on DEV. Apologies for the size of the PR, again its mostly refactoring work than anything new.

NOTE: **Do not forget to expand `PackagesController.cs`,** github seems to hide the large file diffs.

### CODE REVIEW

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 1]
```diff
@@ -185,41 +188,91 @@ public virtual async Task<ActionResult> UploadPackage()
...
-                    verifyRequest.Warnings.AddRange(validationResult.Warnings);
+        private async Task<ActionResult> UploadSymbolsPackageInternal(SubmitPackageRequest model,
+            Stream uploadStream,
+            PackageArchiveReader packageArchiveReader,
+            PackageMetadata packageMetadata,
+            User currentUser)
+        {
+            var symbolsPackageValidationResult = await _symbolPackageUploadService.ValidateUploadedSymbolsPackage(uploadStream, currentUser);
+            var uploadResult = GetJsonResultOrNull(symbolsPackageValidationResult);
+            if (uploadResult != null)
```
Comment (cristinamanum) [inline src/NuGetGallery/Controllers/PackagesController.cs]: Does this mean that it was an error during validation? If yes maybe add a comment?

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 2]
```diff
@@ -185,41 +188,91 @@ public virtual async Task<ActionResult> UploadPackage()
...
+            {
+                TempData["Message"] = symbolsPackageValidationResult.Message;
+                return View(model);
+            }
 
-                    model.InProgressUpload = verifyRequest;
-                }
+            var packageForUploadingSymbols = symbolsPackageValidationResult.Package;
```
Comment (cristinamanum) [inline src/NuGetGallery/Controllers/PackagesController.cs]: Can the symbolsPackageValidationResult be null?

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 3]
```diff
@@ -185,41 +188,91 @@ public virtual async Task<ActionResult> UploadPackage()
...
+            var existingPackageRegistration = packageForUploadingSymbols.PackageRegistration;
+            var id = existingPackageRegistration.Id;
+            var version = packageForUploadingSymbols.NormalizedVersion;
+
+            IEnumerable<User> accountsAllowedOnBehalfOf = Enumerable.Empty<User>();
+            bool isAllowed = ActionsRequiringPermissions.UploadNewPackageVersion.CheckPermissionsOnBehalfOfAnyAccount(currentUser, existingPackageRegistration, out accountsAllowedOnBehalfOf) == PermissionsCheckResult.Allowed;
```
Comment (cristinamanum) [inline src/NuGetGallery/Controllers/PackagesController.cs]: nit: shorter line

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 4]
```diff
@@ -185,41 +188,91 @@ public virtual async Task<ActionResult> UploadPackage()
...
+            if (!isAllowed)
+            {
```
Comment (cristinamanum) [inline src/NuGetGallery/Controllers/PackagesController.cs]: If not allowed why set it as allowed?

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 5]
```diff
@@ -185,41 +188,91 @@ public virtual async Task<ActionResult> UploadPackage()
...
+                TempData["Message"] = symbolsPackageValidationResult.Message;
+                return View(model);
+            }
 
-                    model.InProgressUpload = verifyRequest;
-                }
+            var packageForUploadingSymbols = symbolsPackageValidationResult.Package;
+            var existingPackageRegistration = packageForUploadingSymbols.PackageRegistration;
+            var id = existingPackageRegistration.Id;
+            var version = packageForUploadingSymbols.NormalizedVersion;
```
Comment (cristinamanum) [inline src/NuGetGallery/Controllers/PackagesController.cs]: Is the version used anywhere?

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 6]
```diff
@@ -285,6 +345,43 @@ public virtual async Task<JsonResult> UploadPackage(HttpPostedFileBase uploadFil
             }
         }
 
+        private async Task<JsonResult> UploadSymbolsPackageInternal(PackageArchiveReader packageArchiveReader, Stream uploadStream, NuspecReader nuspec, PackageMetadata packageMetadata)
+        {
+            var currentUser = GetCurrentUser();
+
```
Comment (cristinamanum) [inline src/NuGetGallery/Controllers/PackagesController.cs]: Can the user be null?

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 7]
```diff
@@ -285,6 +345,43 @@ public virtual async Task<JsonResult> UploadPackage(HttpPostedFileBase uploadFil
...
+            var packageForUploadingSymbols = symbolsPackageValidationResult.Package;
+            var existingPackageRegistration = packageForUploadingSymbols.PackageRegistration;
+            var id = existingPackageRegistration.Id;
+            var version = packageForUploadingSymbols.NormalizedVersion;
+
+            // Evaluate the permissions for user on behalf of any account possible, since the user 
+            // could change the ownership before submitting the package.
+            if (ActionsRequiringPermissions.UploadNewPackageVersion.CheckPermissionsOnBehalfOfAnyAccount(
+                currentUser, existingPackageRegistration, out accountsAllowedOnBehalfOf) != PermissionsCheckResult.Allowed)
```
Comment (cristinamanum) [inline src/NuGetGallery/Controllers/PackagesController.cs]: If the user is not allowed to submit why we return the Id not available? Is this the single way for the user not being allowed to submit?

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 8]
```diff
@@ -285,6 +345,43 @@ public virtual async Task<JsonResult> UploadPackage(HttpPostedFileBase uploadFil
...
+            var symbolsPackageValidationResult = await _symbolPackageUploadService.ValidateUploadedSymbolsPackage(uploadStream, currentUser);
+            var uploadResult = GetJsonResultOrNull(symbolsPackageValidationResult);
+            if (uploadResult != null)
+            {
+                return uploadResult;
+            }
+
+            var packageForUploadingSymbols = symbolsPackageValidationResult.Package;
+            var existingPackageRegistration = packageForUploadingSymbols.PackageRegistration;
+            var id = existingPackageRegistration.Id;
```
Comment (cristinamanum) [inline src/NuGetGallery/Controllers/PackagesController.cs]: Are id and version used anywhere?

File: src/NuGetGallery/Scripts/gallery/async-file-upload.js [hunk 9]
```diff
@@ -290,7 +290,9 @@
                 window.nuget.configureExpanderHeading("submit-package-form");
             }
 
-            bindReadMeData(model);
+            if (model === null || !model.IsSymbolsPackage) {
+                bindReadMeData(model);
```
Comment (cristinamanum) [inline src/NuGetGallery/Scripts/gallery/async-file-upload.js]: Will this call be ok on null model?

File: src/NuGetGallery/Controllers/PackagesController.cs (hunk 1 above)
Comment (shishirx34) [reply src/NuGetGallery/Controllers/PackagesController.cs]: This is the same as that of Package synchronous validations, I think it is self explanatory. If you can suggest better method or variable names to make it readable, I am open.

File: src/NuGetGallery/Controllers/PackagesController.cs (hunk 2 above)
Comment (shishirx34) [reply src/NuGetGallery/Controllers/PackagesController.cs]: No. Also, the `.Package` will not be null if code reaches here.

File: src/NuGetGallery/Controllers/PackagesController.cs (hunk 4 above)
Comment (shishirx34) [reply src/NuGetGallery/Controllers/PackagesController.cs]: The above checks if `currentUser` is allowed to push as anyone else(org?). `CheckPermissionsOnBehalfOfAnyAccount`. If the user is not allowed to upload as anyone else, then the only allowed list of owners is the `currentUser`.

File: src/NuGetGallery/Controllers/PackagesController.cs (hunk 5 above)
Comment (shishirx34) [reply src/NuGetGallery/Controllers/PackagesController.cs]: Yes. https://github.com/NuGet/NuGetGallery/blob/symbols-gallery-upload/src/NuGetGallery/Services/SymbolPackageUploadService.cs#L81

Comment (shishirx34) [reply src/NuGetGallery/Controllers/PackagesController.cs]: id/version not used, removed it.

File: src/NuGetGallery/Controllers/PackagesController.cs (hunk 6 above)
Comment (shishirx34) [reply src/NuGetGallery/Controllers/PackagesController.cs]: nope.. we have `UIAuthorize` attribute on the entry points to ensure.

File: src/NuGetGallery/Controllers/PackagesController.cs (hunk 7 above)
Comment (shishirx34) [reply src/NuGetGallery/Controllers/PackagesController.cs]: This checks if the user is allowed to push to an existing package for a new version. If this returns as not allowed it means that the package is owned by someone else and this user is trying to push to a package which already exists. Hence Id not available.

File: src/NuGetGallery/Scripts/gallery/async-file-upload.js (hunk 9 above)
Comment (shishirx34) [reply src/NuGetGallery/Scripts/gallery/async-file-upload.js]: Yes

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 10]
```diff
@@ -1497,253 +1604,381 @@ public virtual async Task<JsonResult> VerifyPackage(VerifyPackageRequest formDat
...
+                            packageArchiveReader,
+                            packageMetadata,
+                            currentUser,
+                            owner);
                     }
+                }
+            }
+            catch (Exception ex)
+            {
+                _telemetryService.TrackPackagePushFailureEvent(id: null, version: null);
```
[the remark this answers is not in the document]
Comment (shishirx34) [reply src/NuGetGallery/Controllers/PackagesController.cs]: Yes nulls are handled correctly, this is an exception case. Shouldn't happen ideally, while unlikely it is possible to happen. I added this check to satisfy a test case which was verifying if unknown exception logs an event with null Id/version. And yes, for symbols and packages there will be additional null events for packages, but I think its very less likely so fine to ignore.

**Comment by shishirx34** [comment]
validated the new changes in DEV env. The Functional and E2E tests passed just fine. new changes look good, except for one minor issue that I have fixed in recent commit.

**Pull request merged**